### ISSUE TITLE
Content of some properties are not being imported using v. 10.4.0 from v. 2.1.6 export

### ISSUE THREAD

**Issue by aguyfromdenmark** [umbraco/Umbraco.Deploy.Issues#234]
When importing content fra an export made using Deploy 2.1.6 [(following this method)](https://docs.umbraco.com/umbraco-deploy/deployment-workflow/import-export/import-export-v7) is imported into a site running Deploy 10.4.0, the content of some properties are not being imported. This happens on nested Nested Content properties (transformed to Block List), and RelatedLinks (transformed to MultiUrlPicker).

For Umbraco HQ, this is regarding ticket #74004.

Reproduction
------------

### Specifics
In a project running Umbraco CMS 7.15.11 and Deploy 2.1.6 with Contrib.Connector and Contrib.Export (2.0.0.7 DLL version), and export to uda files is made using the method found [here](https://docs.umbraco.com/umbraco-deploy/deployment-workflow/import-export/import-export-v7#exporting-umbraco-7-content-and-schema). The content is the imported into a site running Umbraco CMS 10.8.6 and Deploy 10.4.0 using the method found [here ](https://docs.umbraco.com/umbraco-deploy/deployment-workflow/import-export/import-export-v7#importing-umbraco-7-content-and-schema)and [here](https://docs.umbraco.com/umbraco-deploy/deployment-workflow/import-export#migrating-whilst-importing). No error is raised, but some content is not imported.

The v 7 site contains a lot of Nested Content properties, which contains Nested Content properties. The content of the nested Nested Content properties are not being imported in the v 10 site. They are exported from the v 7 site however. 

All the migrators in the v 10 site are kept in one file like this. Different orders of appending has been experiemented with, but they all lead to the same result
```
internal class Migrators : IComposer
{
    public void Compose(IUmbracoBuilder builder)
    {
        builder.DeployArtifactTypeResolvers()
            .AddLegacyTypeResolver();

        builder.DeployPropertyTypeMigrators()
           .Append<NestedContentPropertyTypeMigrator>()
           .Append<MediaPickerPropertyTypeMigrator>();

        builder.DeployArtifactMigrators()
            .Append<ReplaceNestedContentDataTypeArtifactMigrator>()
            .Append<ReplaceMediaPickerDataTypeArtifactMigrator>()
            .AddLegacyMigrators()
            .Append<ElementTypeArtifactMigrator>()
            .Append<FormArtifactMigrator>();


    }

    private class ElementTypeArtifactMigrator : ElementTypeArtifactMigratorBase
    {
        public ElementTypeArtifactMigrator()
            : base(new string[] {
            "foo",
            "bar",
            })
        {
        }
    }

    private sealed class FormArtifactMigrator : ArtifactMigratorBase<FormArtifact>
    {
        protected override FormArtifact? Migrate(FormArtifact artifact)
        {
            // Populate all AllowedUploadTypes with an empty array if it's null to avoid a null reference exception 
            foreach (FormArtifact.Field field in artifact.Pages.SelectMany(x => x.FieldSets.SelectMany(x => x.Containers.SelectMany(x => x.Fields))))
            {
                field.AllowedUploadTypes ??= Array.Empty<FormArtifact.AllowedUploadType>();
            }

            return artifact;
        }
    }
}
```

### Expected result

I have done a export/import migration like this with v 8 to v 10 sites, also with Nested Content properties containing nested Nested Content properties. This has worked without any flaws.

### Actual result

The content is not imported into the v 10 site. 

---
_This item has been added to our backlog AB#43711_

**Comment by ronaldbarendse** [comment]
Hi Andreas, thanks for the detailed description! I've been able to reproduce the issue when using the 'legacy' export in CMS v7 and can confirm this is due to the way the Nested Content properties are stored/exported in this version: the nested values are not strings, but JSON values.

We did introduce a workaround in the Deploy versions that officially introduced this feature (4.9, 10.3 and 12.1+), ensuring nested values are always exported as strings (because [value connectors](https://docs.umbraco.com/umbraco-deploy/getting-started/extending#value-connectors) expect artifact property values to be strings).

Instead of backporting this workaround, I've created a fix for Deploy 13+ to 'parse' the nested JSON property values back to strings, so they can be correctly handled by the value converters and subsequently migrated to the Block List.

We're already planning on releasing a new patch version, so expect the fix the be available within a week 💪🏻

**Comment by aguyfromdenmark** [comment]
Thanks @ronaldbarendse ! That is great to hear! :)

**Comment by ronaldbarendse** [comment]
I've just released patch version 13.2.1 (and 14.1.2) to NuGet that fixes this issue, so you can manually upgrade your [Cloud](https://www.nuget.org/packages/Umbraco.Deploy.Cloud/13.2.1) or [OnPrem](https://www.nuget.org/packages/Umbraco.Deploy.OnPrem/13.2.1) project.

**Comment by ronaldbarendse** [comment]
> @ronaldbarendse What about a patch for v10?

Version 10 is currently in security phase (Deploy follows the [CMS LTS and EOL policy](https://umbraco.com/products/knowledge-center/long-term-support-and-end-of-life/)), so we're not actively releasing bugfixes anymore. This version will also be EOL within a year (June 2025) and the .NET 6 runtime in about 2 months (November 2024, see [.NET official support policy](https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core)).

Have you considered migrating directly to version 13? The changes between 10 and 13 are minimal and you get a solid 2 years of support before it's EOL in December 2026... If this really isn't an option (as in: you have a very convincing reason), we might be able to backport the fix and still release a patch for v10, but would be some time next week at the earliest.

**Comment by aguyfromdenmark** [comment]
@ronaldbarendse I just upgraded the site to v13. Installed the patch. Cleared the content. Imported it again along with schemas. All nodes now throw exceptions regarding JSON parsing when I try to open them. I have notified your support colleagues about it on the ticket #74004.

**Comment by ronaldbarendse** [comment]
> @ronaldbarendse I just upgraded the site to v13. Installed the patch. Cleared the content. Imported it again along with schemas. All nodes now throw exceptions regarding JSON parsing when I try to open them. I have notified your support colleagues about it on the ticket #74004.

Thanks for following up with support and providing the error/stack trace and export ZIP archive to them. With the above issue fixed and nested/recursive Nested Content property values now correctly getting imported, another issue with legacy values in those nested values has surfaced 😞

This time it was something we needed to fix in Deploy Contrib, so have a look at all the details and changes in PR https://github.com/umbraco/Umbraco.Deploy.Contrib/pull/71.

**Comment by aguyfromdenmark** [comment]
@ronaldbarendse That looks really good. Any ETA on the PR being merged and a new package released?

**Comment by ronaldbarendse** [comment]
> @ronaldbarendse That looks really good. Any ETA on the PR being merged and a new package released?

We're in the process of reviewing it now, but once the PR is merged, we don't have any other fixes/features planned and can release it straight after that 👍🏻 I did push the preview version to our nightly MyGet feed, so you can (more easily) test this out already (please provide any feedback on the PR itself): https://www.myget.org/feed/umbraconightly/package/nuget/Umbraco.Deploy.Contrib/13.3.0--alpha.preview.7.gafeddf7.

**Comment by ClientiWeb** [comment]
@ronaldbarendse I downloaded the nightly build and implemented it like in your PR. It works.

**Comment by aguyfromdenmark** [comment]
@ronaldbarendse I downloaded the nightly build and implemented it like in your PR. It works.

**Comment by ronaldbarendse** [comment]
> @ronaldbarendse I downloaded the nightly build and implemented it like in your PR. It works.

Awesome, thanks for testing it out on your import 💪🏻

The fix is now publicly available on NuGet in Deploy Contrib [13.3.0](https://www.nuget.org/packages/Umbraco.Deploy.Contrib/13.3.0) and [14.2.0](https://www.nuget.org/packages/Umbraco.Deploy.Contrib/14.2.0).
